lib/hooks/index.js

```diff
@@ -136,33 +132,33 @@ function activate(agent) {
     // time that we need to go and patch pro-actively.
     for (var moduleName in toInstrument) {
       var instrumentation = toInstrument[moduleName];
-      if (instrumentation.module) {
-        loadHookAndPatch(originalModuleLoad, instrumentation);
+      for (var moduleRoot in instrumentation.patches) {
+        var modulePatch = instrumentation.patches[moduleRoot];
+        if (modulePatch) {
+          loadHookAndPatch(originalModuleLoad, instrumentation, moduleRoot, null);
+        }
       }
     }
 
     // Future requires get patched as they get loaded.
     return function Module_load() {
       var request = arguments[0];
-      var loaded = originalModuleLoad.apply(this, arguments);
-
-      if (typeof request === 'string') {
-        var instrumentation = toInstrument[request];
-
-        if (instrumentation && agent.config().excludedHooks &&
-            agent.config().excludedHooks.indexOf(request) === -1 &&
-            !instrumentation.hook) { // not already patched.
-          instrumentation.version = findModuleVersion(request, arguments[1],
-              originalModuleLoad);
-          logger.info('Patching ' + request + ' at version ' +
-              instrumentation.version);
-          instrumentation.module = loaded;
-          loadHookAndPatch(originalModuleLoad, instrumentation);
+      var instrumentation = toInstrument[request];
+
+      if (instrumentation &&
+          agent.config().excludedHooks.indexOf(request) === -1) {
+        var moduleRoot = findModulePath(request, arguments[1]);
```

Comment:
Let's start using named arguments. The module API is locked, so it ought to be safe.